Reducing build verbosity - pass 1

Leaving aside the question of a `verbose` flag that reduces output further still...

I'm thinking about making two changes.

One: instead of printing full path and every file when parsing as we currently do, I think we might instead print something like this

```
Compiling Grammar: silver:langutil
	At [/home/tedinski/repos/silver/grammars/silver/langutil/]
	Found [Location.sv Attributes.sv DocConfig.sv Message.sv]
```

The downside to this is that it will make it more difficult to open these files. (Right now, you can double click the name and the whole path gets highlighted, makes things quite easy...) But I think we can improve the error messages to include the full path to the file, solving that issue.

Two: I don't know that there's any reason why we have separate passes for generating interface files and generating other translation. I intend to unify them into one.

I think I'm going to try to change the output to look like this:

```
Compiling silver:langutil [/home/tedinski/repos/silver/grammars/silver/langutil/]
	[Location.sv Attributes.sv DocConfig.sv Message.sv]
Found core [/home/tedinski/repos/silver/generated/src/core/Silver.svi]
```

And I'm going to change error message output as well, to better give copy & pastable paths.

Parse errors need serious attention, because they currently forget to report grammar! But also normal errors. I propose something like this:

```
Errors for silver:definition:core
[/home/tedinski/repos/silver/grammars/silver/definition/core/AttributeDcl.sv]
AttributeDcl.sv:19:43: error: Undeclared value 'fName'.

[/home/tedinski/repos/silver/grammars/silver/definition/core/NonTerminalDcl.sv]
NonterminalDcl.sv:19:43: error: Undeclared value 'fName'.

Errors for silver:extension:convenience
[/home/tedinski/repos/silver/grammars/silver/extension/convenience/Convenience.sv]
Convenience.sv:46:4: error: Undeclared value 'attributeDclInh'.
```

That is, before each file's errors we print the full path, to make it easier to open the file.

I like it.

On Wed, May 17, 2017 at 12:39 PM, Ted Kaminski <notifications@github.com>
wrote:

> I think I'm going to try to change the output to look like this:
>
> Compiling silver:langutil [/home/tedinski/repos/silver/grammars/silver/langutil/]
> 	[Location.sv Attributes.sv DocConfig.sv Message.sv]
> Found core [/home/tedinski/repos/silver/generated/src/core/Silver.svi]
>
> And I'm going to change error message output as well, to better give copy
> & pastable paths.
>
> Parse errors need serious attention, because they currently forget to
> report grammar! But also normal errors. I propose something like this:
>
> Errors for silver:definition:core
> [/home/tedinski/repos/silver/grammars/silver/definition/core/AttributeDcl.sv]
> AttributeDcl.sv:19:43: error: Undeclared value 'fName'.
>
> [/home/tedinski/repos/silver/grammars/silver/definition/core/NonTerminalDcl.sv]
> NonterminalDcl.sv:19:43: error: Undeclared value 'fName'.
>
> Errors for silver:extension:convenience
> [/home/tedinski/repos/silver/grammars/silver/extension/convenience/Convenience.sv]
> Convenience.sv:46:4: error: Undeclared value 'attributeDclInh'.
>
> That is, before each file's errors we print the full path, to make it
> easier to open the file.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/melt-umn/silver/issues/151#issuecomment-302169753>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADOhd2etj95t28N_DWU9xaMdTM2myPr2ks5r6zDlgaJpZM4NJhRT>
> .
>

--- a/grammars/silver/composed/idetest/Analyze.sv
+++ b/grammars/silver/composed/idetest/Analyze.sv
@@ -106,8 +106,8 @@ function getIdeMessages
 {
   return map(rewriteMessage(path, _), 
     if !null(spec.parsingErrors)
-    then spec.parsingErrors
-    else spec.errors);
+    then flatMap(snd, spec.parsingErrors)
+    else flatMap(snd, spec.grammarErrors));
 }
 
 function rewriteMessage

--- a/grammars/silver/definition/core/GrammarParts.sv
+++ b/grammars/silver/definition/core/GrammarParts.sv
@@ -7,7 +7,7 @@ nonterminal Grammar with
   grammarName, env, globalImports, grammarDependencies,
   -- Synthesized attributes
   declaredName, moduleNames, exportedGrammars, optionalGrammars, condBuild,
-  defs, importedDefs, errors;
+  defs, importedDefs, grammarErrors;
 
 {--
 - A list of grammars that this grammar depends upon,
@@ -26,6 +26,10 @@ autocopy attribute globalImports :: Decorated Env;
  - At the top of a grammar, these are echoed down as globalImports
  -}
 synthesized attribute importedDefs :: [Def];
+{--
+ - An overall listing of error messages for a grammar
+ -}
+synthesized attribute grammarErrors :: [Pair<String [Message]>];
 
 abstract production nilGrammar
 top::Grammar ::=
@@ -40,7 +44,7 @@ top::Grammar ::=
   
   top.importedDefs = [];
   top.defs = [];
-  top.errors := [];
+  top.grammarErrors = [];
 }
 
 abstract production consGrammar
@@ -54,6 +58,8 @@ top::Grammar ::= h::Root  t::Grammar
 
   top.importedDefs = h.importedDefs ++ t.importedDefs;
   top.defs = h.defs ++ t.defs;
-  top.errors := h.errors ++ t.errors;
+  top.grammarErrors =
+    if null(h.errors) then t.grammarErrors
+    else pair(h.location.filename, h.errors) :: t.grammarErrors;
 
 }

--- a/grammars/silver/definition/core/QName.sv
+++ b/grammars/silver/definition/core/QName.sv
@@ -82,7 +82,7 @@ function dclinfo2possibility
 String ::= dcl::DclInfo
 {
   -- TODO: perhaps some way of including types, when they are relevant (attributes, values)
-  return "\t" ++ dcl.fullName ++ " (" ++ dcl.sourceLocation.filename ++ ":" ++ toString(dcl.sourceLocation.line) ++ ")\n";
+  return "\t" ++ dcl.fullName ++ " (" ++ dcl.sourceLocation.filename ++ ":" ++ toString(dcl.sourceLocation.line) ++ ")";
 }
 
 

--- a/grammars/silver/driver/CompileFiles.sv
+++ b/grammars/silver/driver/CompileFiles.sv
@@ -15,7 +15,7 @@ IOVal<Pair<[Root] [ParseError]>> ::= svParser::SVParser  gpath::String  files::[
   
   -- Print the path we're reading, and read the file.
   local text :: IOVal<String> =
-    readFile(gpath ++ file, print("\t[" ++ gpath ++ file ++ "]\n", ioin));
+    readFile(gpath ++ file, ioin);
 
   -- This is where a .sv file actually gets parsed:
   local r :: ParseResult<Root> = svParser(text.iovalue, file);

--- a/grammars/silver/driver/CompileGrammar.sv
+++ b/grammars/silver/driver/CompileGrammar.sv
@@ -12,7 +12,7 @@ IOVal<Maybe<RootSpec>> ::=
   clean::Boolean
   ioin::IO
 {
-  --the grammar path ':' replaced by '/'
+  -- the grammar path ':' replaced by '/'
   local gramPath :: String = grammarToPath(grammarName);
 
   -- the location (if found) of the grammar
@@ -25,12 +25,13 @@ IOVal<Maybe<RootSpec>> ::=
   local ifaceTime :: IOVal<Maybe<Integer>> = isValidInterface(benv.silverGen ++ "src/" ++ gramPath ++ "Silver.svi", files.io);
   local grammarTime :: IOVal<Integer> = fileTimes(grammarLocation.iovalue.fromJust, files.iovalue, ifaceTime.io);
 
-  local pr :: IO = print("Compiling Grammar: " ++ grammarName ++ "\n", grammarTime.io);
-  
+  -- Decide to compile a grammar or find an interface file
+  local pr :: IO = print("Compiling " ++ grammarName ++ "\n\t[" ++ grammarLocation.iovalue.fromJust ++ "]\n\t[" ++ renderFileNames(files.iovalue, 0) ++ "]\n", grammarTime.io);
   local gramCompile :: IOVal<Pair<[Root] [ParseError]>> =
     compileFiles(svParser, grammarLocation.iovalue.fromJust, files.iovalue, pr);
+
   local ifaceCompile :: IOVal<ParseResult<IRoot>> =
-    compileInterface(sviParser, benv.silverGen ++ "src/" ++ gramPath, pr);
+    compileInterface(sviParser, grammarName, benv.silverGen ++ "src/" ++ gramPath, grammarTime.io);
   
   -- Not being clean, valid interface file, newer than the grammar source
   local useInterface :: Boolean = !clean && ifaceTime.iovalue.isJust && ifaceTime.iovalue.fromJust > grammarTime.iovalue;
@@ -40,8 +41,11 @@ IOVal<Maybe<RootSpec>> ::=
       if ifaceCompile.iovalue.parseSuccess then
         ifaceCompile.io
       else
-        print("\n\tFailed to parse interface file!\n" ++ ifaceCompile.iovalue.parseErrors ++
-                 "\n\tRecovering by parsing grammar....\n", gramCompile.io)
+        -- Do something weird. Demand interface parser finishes, print.
+        -- Then return the normal compile io token branch... Hey, works!
+        unsafeTrace(gramCompile.io,
+          print("\n\tFailed to parse interface file!\n" ++ ifaceCompile.iovalue.parseErrors ++
+                "\n\tRecovering by parsing grammar....\n", ifaceCompile.io))
     else gramCompile.io;
   
   local rs :: RootSpec =
@@ -107,5 +111,16 @@ IOVal<Integer> ::= dir::String is::[String] i::IO
          else if ft.iovalue > rest.iovalue
               then ioval(rest.io, ft.iovalue)
               else rest;
+}
+
+-- A crude approximation of line wrapping
+function renderFileNames
+String ::= files::[String]  depth::Integer
+{
+  return
+    if null(files) then "" else
+    if depth >= 9 then "\n\t " ++ renderFileNames(files, 0) else
+    head(files) ++
+    if null(tail(files)) then "" else " " ++ renderFileNames(tail(files), depth + 1);
 }
 

--- a/grammars/silver/driver/CompileInterface.sv
+++ b/grammars/silver/driver/CompileInterface.sv
@@ -2,12 +2,15 @@ grammar silver:driver;
 
 
 function compileInterface
-IOVal<ParseResult<IRoot>> ::= sviParser::SVIParser  genPath::String  ioin::IO
+IOVal<ParseResult<IRoot>> ::= sviParser::SVIParser  grammarName::String  genPath::String  ioin::IO
 {
   local file :: String = "Silver.svi";
   
+  local pr :: IO = 
+    print("Found " ++ grammarName ++ "\n\t[" ++ genPath ++ file ++ "]\n", ioin);
+
   local text :: IOVal<String> =
-    readFile(genPath ++ file, print("\t[" ++ genPath ++ file ++ "]\n", ioin));
+    readFile(genPath ++ file, pr);
 
   local ir :: ParseResult<IRoot> = sviParser(text.iovalue, file);
 

--- a/grammars/silver/driver/GrammarAction.sv
+++ b/grammars/silver/driver/GrammarAction.sv
@@ -44,6 +44,15 @@ String ::= r::Decorated RootSpec genPath::String
 abstract production printAllBindingErrors
 top::DriverAction ::= specs::[Decorated RootSpec]
 {
+  -- Force printing of status before doing error checks
+  top.code = unsafeTrace(forward.code, forward.ioIn);
+  -- For anyone encountering this hack for the first time,
+  -- IO-token passing can force linearity of actions, but
+  -- interleaves pure computation in annoying ways.
+  -- Without the above, all the error checking work gets done
+  -- (to compute return code) before something tries to do IO,
+  -- so we wouldn't print first.
+
   forwards to printAllBindingErrorsHelp(specs)
   with {
     ioIn = print("Checking For Errors.\n", top.ioIn);

--- a/grammars/silver/driver/GrammarAction.sv
+++ b/grammars/silver/driver/GrammarAction.sv
@@ -62,12 +62,12 @@ top::DriverAction ::= specs::[Decorated RootSpec]
 abstract production printAllBindingErrorsHelp
 top::DriverAction ::= specs::[Decorated RootSpec]
 {
-  local es :: [Message] = head(specs).errors;
+  local es :: [Pair<String [Message]>] = head(specs).grammarErrors;
 
   local i :: IO =
     if null(es)
     then top.ioIn
-    else print("Errors for : " ++ head(specs).declaredName ++ " :\n" ++ foldMessages(es) ++ "\n\n", top.ioIn);
+    else print("Errors for " ++ head(specs).declaredName ++ "\n" ++ sflatMap(renderMessages(head(specs).grammarSource, _), es) ++ "\n", top.ioIn);
 
   local recurse :: DriverAction = printAllBindingErrorsHelp(tail(specs));
   recurse.ioIn = i;
@@ -75,7 +75,7 @@ top::DriverAction ::= specs::[Decorated RootSpec]
   top.io = if null(specs) then top.ioIn else recurse.io;
 
   top.code = 
-    if null(specs) || (!containsErrors(es, head(specs).config.warnError) && recurse.code == 0)
+    if null(specs) || (!grammarContainsErrors(es, head(specs).config.warnError) && recurse.code == 0)
     then 0
     else 20;
 
@@ -90,5 +90,18 @@ top::DriverAction ::= specs::[Decorated RootSpec]
   top.io = if null(errs) then top.ioIn else print(foldMessages(errs), top.ioIn);
   top.order = 0;
   top.code = if null(errs) then 0 else 21;
+}
+
+function renderMessages
+String ::= grammarSource::String  msg::Pair<String [Message]>
+{
+  return " [" ++ grammarSource ++ msg.fst ++ "]\n" ++ foldMessages(msg.snd);
+}
+
+function grammarContainsErrors
+Boolean ::= es::[Pair<String [Message]>]  werr::Boolean
+{
+  return if null(es) then false
+  else containsErrors(head(es).snd, werr) || grammarContainsErrors(tail(es), werr);
 }
 

--- a/grammars/silver/driver/GrammarAction.sv
+++ b/grammars/silver/driver/GrammarAction.sv
@@ -62,12 +62,12 @@ top::DriverAction ::= specs::[Decorated RootSpec]
 abstract production printAllBindingErrorsHelp
 top::DriverAction ::= specs::[Decorated RootSpec]
 {
-  local es :: [Pair<String [Message]>] = head(specs).grammarErrors;
+  local errs :: [Pair<String [Message]>] = head(specs).grammarErrors;
 
   local i :: IO =
-    if null(es)
+    if null(errs)
     then top.ioIn
-    else print("Errors for " ++ head(specs).declaredName ++ "\n" ++ sflatMap(renderMessages(head(specs).grammarSource, _), es) ++ "\n", top.ioIn);
+    else print("Errors for " ++ head(specs).declaredName ++ "\n" ++ sflatMap(renderMessages(head(specs).grammarSource, _), errs) ++ "\n", top.ioIn);
 
   local recurse :: DriverAction = printAllBindingErrorsHelp(tail(specs));
   recurse.ioIn = i;
@@ -75,7 +75,7 @@ top::DriverAction ::= specs::[Decorated RootSpec]
   top.io = if null(specs) then top.ioIn else recurse.io;
 
   top.code = 
-    if null(specs) || (!grammarContainsErrors(es, head(specs).config.warnError) && recurse.code == 0)
+    if null(specs) || (!grammarContainsErrors(errs, head(specs).config.warnError) && recurse.code == 0)
     then 0
     else 20;
 
@@ -85,11 +85,24 @@ top::DriverAction ::= specs::[Decorated RootSpec]
 abstract production printAllParsingErrors
 top::DriverAction ::= specs::[Decorated RootSpec]
 {
-  local errs :: [Message] = flatMap((.parsingErrors), specs);
+  local errs :: [Pair<String [Message]>] = head(specs).parsingErrors;
 
-  top.io = if null(errs) then top.ioIn else print(foldMessages(errs), top.ioIn);
+  local i :: IO =
+    if null(errs)
+    then top.ioIn
+    else print("Errors for " ++ head(specs).declaredName ++ "\n" ++ sflatMap(renderMessages(head(specs).grammarSource, _), errs) ++ "\n", top.ioIn);
+
+  local recurse :: DriverAction = printAllParsingErrors(tail(specs));
+  recurse.ioIn = i;
+
+  top.io = if null(specs) then top.ioIn else recurse.io;
+
+  top.code = 
+    if null(specs) || recurse.code == 0
+    then 0
+    else 21;
+
   top.order = 0;
-  top.code = if null(errs) then 0 else 21;
 }
 
 function renderMessages

--- a/grammars/silver/driver/util/RootSpec.sv
+++ b/grammars/silver/driver/util/RootSpec.sv
@@ -23,7 +23,7 @@ synthesized attribute translateGrammars :: [Decorated RootSpec];
 {--
  - Parse errors present in this grammar (only for errorRootSpec!)
  -}
-synthesized attribute parsingErrors :: [Message];
+synthesized attribute parsingErrors :: [Pair<String [Message]>];
 
 {--
  - Create a RootSpec from a real grammar, a set of .sv files.
@@ -126,15 +126,17 @@ top::RootSpec ::= e::[ParseError]  grammarName::String  grammarSource::String  g
 }
 
 function parseErrorToMessage
-Message ::= grammarSource::String  e::ParseError
+Pair<String [Message]> ::= grammarSource::String  e::ParseError
 {
   return case e of
   | syntaxError(str, locat, _, _) ->
-      err(loc(locat.filename, locat.line, locat.column, locat.endLine, locat.endColumn, locat.index, locat.endIndex),
-          "Syntax error:\n" ++ str)
+      pair(locat.filename, 
+        [err(locat,
+          "Syntax error:\n" ++ str)])
   | unknownParseError(str, file) ->
-      err(loc(grammarSource ++ file, -1, -1, -1, -1, -1, -1),
-          "Unknown error while parsing:\n" ++ str)
+      pair(file,
+        [err(loc(grammarSource ++ file, -1, -1, -1, -1, -1, -1),
+          "Unknown error while parsing:\n" ++ str)])
   end;
 }
 

--- a/grammars/silver/driver/util/RootSpec.sv
+++ b/grammars/silver/driver/util/RootSpec.sv
@@ -1,6 +1,6 @@
 grammar silver:driver:util;
 
-import silver:definition:core only Grammar, errors, grammarName, importedDefs, grammarDependencies, globalImports, Message, err;
+import silver:definition:core only Grammar, grammarErrors, grammarName, importedDefs, grammarDependencies, globalImports, Message, err;
 import silver:definition:env:env_parser only IRoot;
 import silver:definition:flow:env only flowEnv, flowDefs, fromFlowDefs;
 import silver:definition:flow:ast only nilFlow, consFlow, FlowDef;
@@ -13,7 +13,7 @@ nonterminal RootSpec with
   config, compiledGrammars, productionFlowGraphs, grammarFlowTypes,
   -- synthesized attributes
   declaredName, moduleNames, exportedGrammars, optionalGrammars, condBuild, allGrammarDependencies,
-  defs, errors, grammarSource, grammarTime, interfaceTime, recheckGrammars, translateGrammars, parsingErrors;
+  defs, grammarErrors, grammarSource, grammarTime, interfaceTime, recheckGrammars, translateGrammars, parsingErrors;
 
 {--
  - Grammars that were read from source.
@@ -68,7 +68,7 @@ top::RootSpec ::= g::Grammar  grammarName::String  grammarSource::String  gramma
   top.allGrammarDependencies = actualDependencies;
   
   top.defs = g.defs;
-  top.errors := g.errors;
+  top.grammarErrors = g.grammarErrors;
   top.parsingErrors = [];
 }
 
@@ -96,7 +96,7 @@ top::RootSpec ::= p::IRoot  interfaceTime::Integer
   top.allGrammarDependencies = p.allGrammarDependencies;
 
   top.defs = p.defs;
-  top.errors := []; -- TODO: consider getting grammarName and comparing against declaredName?
+  top.grammarErrors = []; -- TODO: consider getting grammarName and comparing against declaredName?
   top.parsingErrors = [];
 }
 
@@ -121,7 +121,7 @@ top::RootSpec ::= e::[ParseError]  grammarName::String  grammarSource::String  g
   top.allGrammarDependencies = [];
 
   top.defs = [];
-  top.errors := [];
+  top.grammarErrors = [];
   top.parsingErrors = map(parseErrorToMessage(grammarSource, _), e);
 }
 

